### ISSUE TITLE
Identity quote failure when using TPM devices with manufacturer certificates signed by ECC keys

### ISSUE THREAD

**Issue by galmasi** [keylime/keylime#1070]
# Environment
 * OS / version: OpenBMC Yocto "phosphor", Ubuntu 20.04, Ubuntu 22.04
 * Processor architecture: ARM cortex A7, AMD x86_64 (mostly irrelevant)
 * TPM Manufacturer: Nuvoton TPM signed by `NUVO_1111.pem`, but reproducible with swtpm.
 * Keylime version: 6.4.0, but problem is reproducible with current master branch (Aug 8 2022)

# Description 

During the initial identity quote at the start of attestation the function `tpm_main.verify_ek()` is invoked by the keylime tenant to ensure that the TPM device on the agent is genuine. The process involves listing all the manufacturer certificates in the `tpm_cert_store` directory, finding one that has an Issuer that matches the EK cert's, and then running verify().

**When the manufacturer certificate signature algorithm is EC(*) the verify() function call fails, resulting in attestation failure.**

## Expected behavior vs. actual behavior 

* Expected behavior: a Nuvoton device certified by the file keylime/tmp_cert_store/NUVO0_1111.pem should work.
* Actual behavior: said Nuvoton TPM device will never result in successful attestation, because the NUVO_1111.pem file is signed with an ECC key. The following error is generated during the identity quote phase by the tenant:

```
2022-08-09 14:43:15.313 - keylime.tpm - ERROR - verify() takes 4 positional arguments but 5 were given
Traceback (most recent call last):
  File "/usr/local/lib/python3.9/site-packages/keylime-6.4.0-py3.9.egg/keylime/tpm/tpm_main.py", line 857, in verify_ek
    signcert.public_key().verify(
TypeError: verify() takes 4 positional arguments but 5 were given
```

## Steps to reproduce problem
1A. Install keylime attestation on any system with a Nuvoton chip that is certified by the NUVO_1111.pem file.
1B. Alternatively, create a swtpm instance in which the manufacturer cert is signed by an ECC key.
2. Start the keylime agent and let it register. Registration should execute normally.
3. Attempt to start attestation with the `keylime_tenant -c add` command. Expect the error message above to show up.

# Root cause analysis

The problem is with the following sequence in the TPM code (`keylime/tpm/tpm_main.py`):

```
                    signcert.public_key().verify(
                        ek509.signature,
                        ek509.tbs_certificate_bytes,
                        padding.PKCS1v15(),
                        ek509.signature_hash_algorithm,
                    )
```

The call to the verify() method makes an assumption that (including "self") 5 arguments are needed. But on elliptic curve public keys only 4 arguments are expected (there is no padding).

* Abstract method signature for RSA public keys: https://github.com/pyca/cryptography/blob/6b63cd56c746150ff49fdfe38c422523afaf21b8/src/cryptography/hazmat/primitives/asymmetric/rsa.py#L98-L104

* Abstract method signature for ECC public keys: https://github.com/pyca/cryptography/blob/6b63cd56c746150ff49fdfe38c422523afaf21b8/src/cryptography/hazmat/primitives/asymmetric/ec.py#L153-L158

This discrepancy causes the problem.

# Proposed (tentative) fix

Insert an if statement to handle the different verify() method signatures in python cryptography.

**Comment by galmasi** [comment]
Addendum: creating an swtpm instance in which the root CA is signed with ECC keys.

1. The following gist shows the code to generate the manufacturer certificate and private key to be used by swtpm. <script src="https://gist.github.com/galmasi/d54ccef12ba14bfc1170afc36b5abd14.js"></script>. Use the gist to generate the root CA and a private key.

1A. Check your work: the newly generated root CA should be signed with ecdsa-with-SHA256.
```
openssl x509 -in root-ca-cert.crt -noout -text
```

2. Create a modified `swtpm-localca.conf` that points to the just-generated root CA and the private key -- they are called `issuercert` and `signingkey` in swtpm terminology. 

3. Point `swtpm_setup.conf` at the modified `swtpm-localca.conf`

4. Run `swtpm_setup` as usual, point it at the modified `swtpm_setup.conf`.

5. Start a virtual machine with the generated swtpm setup, and run attestation using the virtual TPM.

**Comment by kkaarreell** [comment]
Hi @galmasi 
I am trying to write a test implementing the proposed setup but I am not able to reproduce the reported issue.
https://github.com/RedHat-SP-Security/keylime-tests/pull/175/files
In the test I am generating ECDSA certificate for TPM cert issuer and by extracting EK cert from TPM I can verify it has been used to sign EK cert itself.
```
# tpm2_nvread 0x1c00002 > ekcert-rsa.der
# openssl x509 -inform der -in ekcert-rsa.der -noout -text
Certificate:
    Data:
        Version: 3 (0x2)
        Serial Number: 5 (0x5)
        Signature Algorithm: ecdsa-with-SHA256
        Issuer: O = RH, OU = keylime-tests, CN = keylime
        Validity
            Not Before: Aug 19 14:15:46 2022 GMT
            Not After : Dec 31 23:59:59 9999 GMT
        Subject: CN = unknown
        Subject Public Key Info:
            Public Key Algorithm: rsaEncryption
                RSA Public-Key: (2048 bit)
                Modulus:
                    00:a0:2c:8c:ed:50:40:b0:7a:41:f6:70:c8:b5:28:
                    ce:15:54:89:09:80:81:db:c7:10:fe:50:fa:9c:f6:
                    60:fa:f1:62:e7:84:05:11:b7:88:90:52:92:e6:f6:
                    5d:ac:c7:15:91:e6:79:00:9d:57:50:db:09:6e:57:
                    aa:a4:34:f1:aa:bb:23:b4:18:eb:38:62:c1:8d:ad:
                    69:18:34:52:66:9b:2d:15:21:62:84:3e:55:20:b9:
                    e5:4c:f7:ee:bc:9c:b6:d6:78:84:5f:00:fb:14:73:
                    61:b1:e1:11:9e:06:ba:11:b5:fa:ca:d0:0c:b5:2d:
                    8c:e4:55:fa:10:86:78:34:b2:e6:db:32:9c:73:b5:
                    50:b6:50:74:21:b9:67:03:3c:02:4d:3e:4c:ab:a3:
                    5e:9f:3b:6b:97:a3:ee:5c:93:ac:80:69:20:e4:a2:
                    b7:25:6e:df:fc:c5:07:2e:c6:12:bb:c3:5c:18:27:
                    b8:77:92:06:8d:f0:e3:76:a9:61:42:bc:c2:fc:be:
                    23:6c:5b:b0:6c:92:73:b3:02:f9:5d:dc:27:1a:d5:
                    54:37:f0:93:b8:7d:5f:7e:1a:5b:e7:24:c4:86:c2:
                    4c:c8:21:f6:a4:aa:f7:34:5d:c6:34:ad:71:de:27:
                    b3:03:84:84:d6:0d:db:71:eb:9f:e3:86:60:e7:21:
                    f8:89
                Exponent: 65537 (0x10001)
        X509v3 extensions:
            X509v3 Extended Key Usage: 
                2.23.133.8.1
            X509v3 Subject Alternative Name: critical
                DirName:/2.23.133.2.1=id:00001014/2.23.133.2.2=swtpm/2.23.133.2.3=id:20191023
            X509v3 Basic Constraints: critical
                CA:FALSE
            X509v3 Subject Directory Attributes: 
                0.0...g....1.0...2.0.......
            X509v3 Authority Key Identifier: 
                keyid:7C:69:BA:DD:B7:B5:23:C5:06:20:F2:DC:D2:50:C1:1A:A0:3D:B3:EB

            X509v3 Key Usage: critical
                Key Encipherment
    Signature Algorithm: ecdsa-with-SHA256
         30:45:02:20:19:67:02:d3:1a:5e:2c:8d:42:ca:d2:66:c0:74:
         ec:e0:aa:7a:56:2b:31:d9:57:93:f0:c0:12:5a:fe:b5:7a:14:
         02:21:00:bf:57:98:8f:72:14:ce:87:cb:bf:f8:2a:70:f0:18:
         a3:8a:83:39:92:ab:1b:16:d7:c6:4a:22:b3:24:bc:38:eb
```
However, in my case there is not error regarding invalid argument count, instead `keylime_tenant` reports
```
2022-08-19 14:23:19.127 - keylime.tpm - ERROR - No Root CA matched EK Certificate
2022-08-19 14:23:19.128 - keylime.tenant - WARNING - Invalid EK certificate
2022-08-19 14:23:19.128 - keylime.tenant - ERROR - TPM Quote from cloud agent is invalid for nonce: 4awMYQXowrOqxJu4LSNT
```
which is expected given I have modified issuer's cert but kept Root CA untouched.

Is it necessary to create new Root CA at first (so it can be used to sign issuer's cert) to reproduce the issue or is there something else wrong with my setup?

**Comment by kkaarreell** [comment]
Finally I am able to reproduce it. The error above is caused by https://github.com/keylime/keylime/issues/1080.

**Comment by galmasi** [comment]
My apologies, I [temporarily] dropped the issue. I think there is a relatively simple way to produce this problem by playing with the manufacturing parameters in swtpm.

**Comment by stefanberger** [comment]
This would probably be a candidate for a patch but I am a bit slow at the moment with getting to a test environment:
```
diff --git a/keylime/tpm/tpm_main.py b/keylime/tpm/tpm_main.py
index 847aabf..393044a 100644
--- a/keylime/tpm/tpm_main.py
+++ b/keylime/tpm/tpm_main.py
@@ -17,6 +17,7 @@ from cryptography import x509
 from cryptography.hazmat.backends import default_backend
 from cryptography.hazmat.primitives import serialization as crypto_serialization
 from cryptography.hazmat.primitives.asymmetric import padding
+from cryptography.hazmat.primitives.asymmetric.rsa import RSAPublicKey
 from packaging.version import Version

 from keylime import cmd_exec, config, keylime_logging, secure_mount, tpm_ek_ca
@@ -805,12 +806,19 @@ class tpm(tpm_abstract.AbstractTPM):
                     continue

                 try:
-                    signcert.public_key().verify(
-                        ek509.signature,
-                        ek509.tbs_certificate_bytes,
-                        padding.PKCS1v15(),
-                        ek509.signature_hash_algorithm,
-                    )
+                    if isinstance(signcert.public_key(), RSAPublicKey):
+                        signcert.public_key().verify(
+                            ek509.signature,
+                            ek509.tbs_certificate_bytes,
+                            padding.PKCS1v15(),
+                            ek509.signature_hash_algorithm,
+                        )
+                    else:
+                        signcert.public_key().verify(
+                            ek509.signature,
+                            ek509.tbs_certificate_bytes,
+                            ek509.signature_hash_algorithm,
+                        )
                 except crypto_exceptions.InvalidSignature:
                     continue

```

**Comment by kkaarreell** [comment]
We have a reproducer available in https://github.com/RedHat-SP-Security/keylime-tests/pull/175 so one way to test it could be updating `tmt` test plans to point to that PR repo and branch by modifying `url` and `ref` in `packit-ci.fmf`
```
url: https://github.com/RedHat-SP-Security/keylime-tests.git
ref: ecc
```

**Comment by sergio-correia** [comment]
> This would probably be a candidate for a patch but I am a bit slow at the moment with getting to a test environment:

Yeah, something like this will resolve it. 

I have a follow-up PR to #1084 to also use pyasn1 in this code path, as we will have the same issues when attempting to parse some broken certs of the tpm cert store with python-cryptography. I ended up moving the verify_ek() logic to cert_utils, so that we could test it more easily, and additionally I fixed this issue with ECC keys -- just submitted it  for review as #1156.
